paddle/framework/scope.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+#include "paddle/framework/variable.h"
+#include "paddle/utils/Error.h"
+
+namespace paddle {
+namespace framework {
+
+const static Error AlreadyCreated("Variable has already been created");
+
+/**
+ * Scope is an association of a name to Variable. All variables belong to
+ * `Scope`. You need to specify a scope to run a Net, i.e., `net.Run(&scope)`.
+ * One net can run in different scopes and update different variable in the
+ * scope.
+ */
+class Scope {
+ public:
+  Scope() {}
+
+  explicit Scope(const std::shared_ptr<Scope>& scope) : parent_(scope) {}
+
+  ~Scope() {}
+
+  // Create Variable in this Scope. Return error if Variable already been
+  // created.
+  Error __must_check CreateVariable(const std::string& name);
```

Comment:
> if the variable is already there, users' intention (create the variable) is fulfilled, there is really nothing worth of returning.

This line convinced me, thanks!  I will try to learn more about the `first principle`.